credit-risk/src/app.py

```diff
@@ -1,28 +1,234 @@
+
+
+async def render_customer_page(q: Q):
+
+    del q.page['customer_table']
+
+    row_index = int(q.args.customer_table[0])
+    customer_row = q.app.customer_df.loc[row_index]
+    score = q.app.predictions_df.loc[row_index]
+    approve = bool(score < APPROVAL_THRESHOLD)
+    contribs = q.app.contributions_df.loc[row_index].drop("BiasTerm")
+
+    q.client.selected_customer_id = customer_row["ID"]
+
+    # details
+    q.page.add(
+        'customer_features', 
+        ui.form_card(
+            box='customer_features',
+            items=[
+                ui.table(
+                    name='customer_features',
+                    columns=[
+                        ui.table_column(name="attribute", label="Attribute", sortable=False, searchable=False, max_width='100'),
+                        ui.table_column(name="value", label="Value", sortable=False, searchable=False, max_width='100')
+                    ],
+                    rows=[ui.table_row(name=index, cells=[index, row]) for index, row in customer_row.map(str).iteritems()],
+                    groupable=False,
+                    resettable=False,
+                    multiple=False,
+                    height='760px'
+                )
+            ]
+        )
+    )
+
+    # summary
+    top_feature = contribs.idxmin(axis=1) if approve else contribs.idxmax(axis=1)
+    explanation_data = {
+        'will_or_will_not': 'will' if approve else 'will not',
+        'top_contributing_feature': top_feature,
+        'value_of_top_contributing_feature': str(customer_row[top_feature]),
+        'accept_or_reject': 'approve' if approve else 'reject',
+    }
+    explanation = (
+        "- This customer **{{will_or_will_not}}** most probably settle the next month credit card balance.\n"
+        "- Having a **{{top_contributing_feature}}** of **{{value_of_top_contributing_feature}}** is the top reason for that.\n"
+        "- It's a good idea to **{{accept_or_reject}}** this customer." 
+    )
+    q.page["customer_risk_explanation"] = ui.markdown_card(
+        box='customer_risk_explanation',
+        title='Summary on Customer',
+        content='=' + explanation,
+        data=explanation_data,
+    )
+
+    # shap plot
+    shap_values = list(zip(contribs.index, contribs))
+    shap_values.sort(key=lambda x: x[1])
+    q.page['customer_shap_plot'] = ui.plot_card(
+        box='customer_shap_plot',
+        title="Effectiveness of each attribute on defaulting next payment",
+        data=data(['label', 'value'], rows=shap_values),
+        plot=ui.plot([ui.mark(type='interval', x='=value', x_title='SHAP value', y='=label')])
+    )
+
+    # approve/reject buttons
+    q.page["button_group"] = ui.form_card(
+        box='button_group',
+        items=[
+            ui.buttons(
+                [
+                    ui.button(name='approve_btn', label='Approve', primary=approve),
```

Comment:
Really nice idea of using primary dynamically

Reply:
Thanks but it was like that when I got here.